lib/galaxy/webapps/galaxy/api/plugins.py

```diff
@@ -46,7 +46,7 @@ def show(self, trans, id, **kwargs):
             result = {}
             result["hdas"] = []
             for hda in history.datasets:
-                if registry.get_visualization(trans, id, hda):
+                if registry.get_visualization(trans, id, hda) and not hda.purged and not hda.deleted:
```

Comment:
```suggestion
                if not hda.deleted and registry.get_visualization(trans, id, hda):
```
Don't think you can have purged but not deleted datasets, but in either case it's cheaper to check if they're deleted first.

Reply:
Ah, and we don't want hidden datasets in there either. So maybe iterate over `history.__dataset_contents_iter()` which will exclude deleted and hidden datasets.